Client/Coordinators/BaseCoordinator.swift

```diff
@@ -37,6 +51,9 @@ open class BaseCoordinator: NSObject, Coordinator {
 
     @discardableResult
     func findAndHandle(route: Route) -> Coordinator? {
+        isPerformingFindAndHandle += 1
+        defer { isPerformingFindAndHandle -= 1 }
```

Comment:
@lmarceau @OrlaM One small addition since this was reviewed earlier: this change allows the existing unit tests (like `testRemoveDismissableFromChild()`) to continue working without requiring a dispatch_async etc. It just skips this workaround entirely unless we can see that we're in the midst of recursing within `findAndHandle(route:)`. It will allow synchronous code to continue working as-is with no changes (since we only need to apply this bandaid when calling `add(child:)` while also recursing in the findAndHandle). LMK if any concerns, happy to remove/update this further.

Reply:
We can maybe use instead a `DispatchQueueInterface` instead of directly calling `DispatchQueue.main.async` in prod code. Basically this enables us to remove the async component of unit tests, so `coordinator?.newlyAdded = false` would be called right away since we mock the dispatch queue (I believe this would fix the current issue with that particular test?)

If it does fix it, then we just need to inject a mock dispatch queue interface from the test, and have a default param of `= DispatchQueue.main` in production code. You can search `mainQueue: DispatchQueueInterface = DispatchQueue.main` in our codebase and you'll see different usage of this `DispatchQueueInterface`.

Reply:
I think you're correct and that would fix the unit tests, my only other thought was that this addition might be safer in the client since ideally I don't think we want this workaround applying to all synchronous code. For ex. if you add a child coordinator on the main thread and somewhere later in the callstack (but before the main thread runloop has completed) we call into `findAndHandle()`, that child won't be dismissable (even if `isDismissable` is true) because it will be flagged as `newlyAdded`.

I don't know all of the use-cases / scenarios well enough to know if that's a problem or not, though I can imagine potential scenarios where this workaround perhaps shouldn't apply. Afaics, the only time the fix is really necessary is when we're adding the child while already recursing `findAndHandle`. So this seemed like a useful change to limit any unintended impacts.

I'm fine with either though so LMK if you have a preference either way, happy to update to use `DispatchQueueInterface`.

Reply:
I think it's preferable to avoid changing production code as much as possible for unit tests.  Hence I think I prefer we go for `DispatchQueueInterface` instead of `isPerformingFindAndHandle` if possible 🙏

Reply:
The `DispatchQueueInterface` approach will incur greater impacts to production since it will cause this workaround to be applied in more scenarios than (I think) it necessarily should be, but I'm Ok with either. Also, it's a separate topic but from a broader perspective it's kind of unfortunate that our unit tests run differently on CI than production. Ideally the tests should be exercising the code that is running for users in prod. If the code is specifically written to behave one way on CI and a different way in prod then I think it slightly diminishes their usefulness. In any case though hopefully this workaround won't be in place very long. I've pushed up the requested changes; I think there is still one test in `SceneCoordinatorTests` that may be failing, taking a closer look at that and then this should be ready to merge once reviewed.

Reply:
The DispatchQueueInterface has pretty much no impact on production because it's just concealing DispatchQueue. 
The only real difference between CI and local is that CI is much slower, this has a tendency to expose flaky tests.

Reply:
> has pretty much no impact on production because it's just concealing DispatchQueue.

The `MockDispatchQueue` operates differently than the default queue (`DispatchQueue.main`). The `testRemoveDismissableFromChild()` is good example I think: using the `MockDispatchQueue` means that unit test will pass on CI, because the code executing there is going to happen in a different order than identical code would in production. The same code + assertions running in production would actually fail. It's one of the reasons I was hoping to advocate for instead making this workaround more targeted (which would obviate the need for `DispatchQueueInterface` in BaseCoordinator entirely), and only apply the workaround when we can see we're in the midst of recursing through `findAndHandle(route:)`.

Reply:
That's a fair point but personally I think it's a worthwhile trade off. On one hand  you have code with expectations set that expire after a set a mount of time which can result in flaky tests in slow environments. On the other hand you have stable tests that aren't necessarily testing the exact same behaviour as production code. In the first case, if something occasionally happens in the wrong order (a real issue) that is caught by tests the developer is likely to just re-run the tests because of the prevalence of flaky tests, thus missing the issue. In both cases the real issue is missed. Historically I haven't found unit tests to be very good at picking up intermittent timing issues or race conditions because of this.
This is a very interesting topic tho, with no clear correct answer. I think it would make a great discussion for our Thursday eng meeting. The agenda is in the invite if you want to add a topic.